### PR TITLE
virtio_fs: create dir with winapi in windows guest

### PR DESCRIPTION
**Pull request by xiagao**
New case about 'Create new directory with WinAPI's CreateDirectory', it is only for windows guest and virtiofsd-rs version.
    1. Create directory on shared dir with Winapi command.
    2. Check virtiofsd output to make sure there is no error.
ID: 2102452
Signed-off-by: Xiaoling Gao <xiagao@redhat.com>

### CODE REVIEW

**Comment by xiagao** [comment]
python3 ConfigTest.py --testcase=virtio_fs_share_data..create_dir_winapi --guestname=Win2022 --clone=no --firmware=ovmf --clone=no
 (1/1) Host_RHEL.m9.u1.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.auto.create_dir_winapi.q35: STARTED
 (1/1) Host_RHEL.m9.u1.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.auto.create_dir_winapi.q35: PASS (248.24 s)

**Comment by xiagao** [comment]
As also change parameters of virtiofsd binary,so run another test case to cover it.
 (1/1) Host_RHEL.m8.u7.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.always.xattr_on.flock_on.lock_posix_on.q35: STARTED
 (1/1) Host_RHEL.m8.u7.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.always.xattr_on.flock_on.lock_posix_on.q35: PASS (148.23 s)

**Comment by leidwang** [comment]
Hi @xiagao Could you add more details to commit message?

**Comment by leidwang** [comment]
If this cases only works on virtiofsd-rs version,do we need to check the virtiofsd version before start the test?

**Comment by xiagao** [comment]
@leidwang 
1. commit is updated.
2. there is no need to check,for rhel9, we only use virtiofsd-rs.
thanks.

File: qemu/tests/cfg/virtio_fs_share_data.cfg [hunk 1]
```diff
@@ -81,38 +81,47 @@
...
                 cdrom_cd1= "/home/winutil.iso"
                 git_check_cmd = '"c:\Program Files\Git\bin\git.exe" init %TEMP%\test'
+        - create_dir_winapi:
+            only Windows
+            only default.default.with_cache.auto.default
+            no Host_RHEL.m8
+            fs_binary_extra_options += " --log-level debug"
+            create_dir_winapi_cmd = python2.7 -c "import ctypes; dll=ctypes.windll.kernel32;dll.CreateDirectoryA('%s:\\test_winapi', 0)"
+            vfsd_winapi_log_name = "/tmp/virtiofs_create_dir_winapi_log"
+            vfsd_winapi_log_cmd = "tail -f %s > ${vfsd_log_winapi}"
```
Comment (leidwang) [inline qemu/tests/cfg/virtio_fs_share_data.cfg:91]: It seems that we did not define `vfsd_log_winapi` before referencing it.

**Comment by leidwang** [comment]
And I run the test cases on my system,it can not pass. Could you help to check it?

 (1/1) Host_RHEL.m9.u1.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.auto.create_dir_winapi.q35: ERROR: Command 'cat /tmp/virtiofs_create_dir_winapi_log' failed.\nstdout: b''\nstderr: b'cat: /tmp/virtiofs_create_dir_winapi_log: No such file or directory\n'\nadditional_info: None (132.15 s)

File: qemu/tests/cfg/virtio_fs_share_data.cfg (hunk 1 above)
Comment (xiagao) [reply qemu/tests/cfg/virtio_fs_share_data.cfg:91]: Yes, I modified vfsd_log_winapi's name but forgot update here.
Thanks for reminder.

File: qemu/tests/cfg/virtio_fs_share_data.cfg [hunk 2]
```diff
@@ -81,38 +81,47 @@
                 git_init_cmd = '"c:\Program Files\Git\bin\git.exe" init %s\test'
                 cdrom_cd1= "/home/winutil.iso"
                 git_check_cmd = '"c:\Program Files\Git\bin\git.exe" init %TEMP%\test'
+        - create_dir_winapi:
+            only Windows
+            only default.default.with_cache.auto.default
+            no Host_RHEL.m8
```
[the remark this answers is not in the document]
Comment (xiagao) [reply qemu/tests/cfg/virtio_fs_share_data.cfg:87]: I thinks no Host_RHEL.m8 is ok,do you have any other concern?

[the remark this answers is not in the document]
Comment (leidwang) [reply qemu/tests/cfg/virtio_fs_share_data.cfg:87]: As you said "there is no need to check,for rhel9, we only use virtiofsd-rs.", if so, currently the logic only works for RHEL9? Right?

[the remark this answers is not in the document]
Comment (xiagao) [reply qemu/tests/cfg/virtio_fs_share_data.cfg:87]: and later...

[the remark this answers is not in the document]
Comment (leidwang) [reply qemu/tests/cfg/virtio_fs_share_data.cfg:87]: But `no Host_RHEL.m8` means include rhel7/6/5... and rhel9/10/11...

[the remark this answers is not in the document]
Comment (xiagao) [reply qemu/tests/cfg/virtio_fs_share_data.cfg:87]: required_qemu = [5.2.0,) will unclude rhel7/6/5, this case will support rhel9+ if there is no exception.

[the remark this answers is not in the document]
Comment (leidwang) [reply qemu/tests/cfg/virtio_fs_share_data.cfg:87]: OK,We defined `no RHEL.6 RHEL.7 ` before.

**Comment by leidwang** [comment]
(1/1) Host_RHEL.m9.u1.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.auto.create_dir_winapi.q35: PASS (130.87 s)

LGTM.

**Review by leidwang — APPROVED** [review]

File: qemu/tests/virtio_fs_share_data.py [hunk 3]
```diff
@@ -491,6 +496,28 @@ def is_autoit_finished(session, process_name):
 
                     if status:
                         test.fail("Git init failed with %s" % output)
+                if create_dir_winapi_cmd:
+                    error_context.context("Get virtiofsd log file.", test.log.info)
+                    vfsd_dev = vm.devices.get_by_params({"source": fs_source})[0]
+                    virtiofsd_log = '%s-%s.log' % (vfsd_dev.get_qid(),
+                                                   vfsd_dev.get_param('name'))
+                    logfile = utils_misc.get_log_filename(virtiofsd_log)
+                    os.popen(vfsd_winapi_log_cmd % logfile)
```
[the remark this answers is not in the document]
Comment (YongxueHong) [reply qemu/tests/virtio_fs_share_data.py:505]: And I think that we should not need to echo the context of the virtio-fs log to another file, but just 
open the virtio-fs log file then check the context.

**Comment by xiagao** [comment]
@YongxueHong  I modified as we talk offline, could you help to review again?
 (1/2) Host_RHEL.m9.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.auto.create_dir_winapi.q35: STARTED
 (1/2) Host_RHEL.m9.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.auto.create_dir_winapi.q35: PASS (105.18 s)
 (2/2) Host_RHEL.m9.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.always.xattr_on.flock_off.lock_posix_off.q35: STARTED
 (2/2) Host_RHEL.m9.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.always.xattr_on.flock_off.lock_posix_off.q35: PASS (111.67 s)

File: qemu/tests/cfg/virtio_fs_share_data.cfg (hunk 2 above)
[the remark this answers is not in the document]
Comment (vivianQizhu) [reply qemu/tests/cfg/virtio_fs_share_data.cfg:87]: > OK,We defined `no RHEL.6 RHEL.7 ` before.

I think we don't need it now, they are out of date, `no RHEL.8` is sufficient now.

File: qemu/tests/virtio_fs_share_data.py [hunk 4]
```diff
@@ -491,6 +495,27 @@ def is_autoit_finished(session, process_name):
...
+                                                  vfsd_dev.get_param('name'))
+                    vfd_logfile = utils_misc.get_log_filename(vfd_log_name)
+
+                    error_context.context("Check virtiofsd log.", test.log.info)
+                    pattern = r'Replying ERROR.*header.*OutHeader.*error.*-9'
+                    with open(vfd_logfile, 'r') as f:
+                        lines = f.readlines()
+                    for line in lines:
+                        if re.match(pattern, line, re.I):
+                            test.fail("CreateDirectory cause virtiofsd-rs ERROR reply.")
```
[the remark this answers is not in the document]
Comment (xiagao) [reply qemu/tests/virtio_fs_share_data.py]: Ok, update.
 (1/2) Host_RHEL.m9.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.auto.create_dir_winapi.q35: STARTED
 (1/2) Host_RHEL.m9.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.auto.create_dir_winapi.q35: PASS (110.74 s)
 (2/2) Host_RHEL.m9.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.always.with_writeback.q35: STARTED
 (2/2) Host_RHEL.m9.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.always.with_writeback.q35: PASS (100.74 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0

**Comment by xiagao** [comment]
Thanks, update.
(1/2) Host_RHEL.m9.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.auto.create_dir_winapi.q35: STARTED
(1/2) Host_RHEL.m9.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.auto.create_dir_winapi.q35: PASS (110.74 s)
(2/2) Host_RHEL.m9.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.always.with_writeback.q35: STARTED
(2/2) Host_RHEL.m9.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.always.with_writeback.q35: PASS (100.74 s)
RESULTS : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0

**Review by YongxueHong — APPROVED** [review]
LGTM.

**Pull request merged**